[myuplink] Initial contribution (#17451)
openhab/openhab-addons: bundles/org.openhab.binding.myuplink/README.md

| dataPollingInterval     | no       | Interval (seconds) in which live data values are retrieved from the Easee Cloud API. (default = 60)                                                                     |

## Thing Configuration

It is recommended to use auto discovery which does not require further configuration.
If manual configuration is preferred you need to specify configuration as below.

| Configuration Parameter | Required | Description                                                                                                            |
|-------------------------|----------|------------------------------------------------------------------------------------------------------------------------|
| deviceId                | yes      | The id of the heatpump that will be represented by this thing. Can be retrieved via API call or autodiscovery.         |
| systemId                | no       | The systemId of the heatpump. Only needed for "SmartHomeMode". Can be retrieved via API call or autodiscovery.         |

> ## Channels

Author: it should be resolved now although it is not marked as "outdated"

# myUplink Binding

The myUplink binding is used to get "live data" from from Nibe heat pumps without plugging any custom devices into your heat pump.
This avoids the risk of losing your warranty.
Instead data is retrieved from myUplink.
The myUplink API is the successor of the Nibe Uplink API.
This binding should in general be compatible with all heat pump models that support myUplink.
Read or write access is supported by all channels as exposed by the API.
Write access might only be available with a paid subscription for myUplink.
You will need to create credentials at https://dev.myuplink.com/apps in order to use this binding.

## Supported Things

This binding provides two thing types:

| Thing/Bridge        | Thing Type          | Description                                                                                   |
|---------------------|---------------------|-----------------------------------------------------------------------------------------------|
| bridge              | account             | cloud connection to a myUplink user account                                                   |

> | thing               | genericDevice       | the physical heatpump which is connected to myUplink                                          |

Reviewer: Please adjust Thing id to the naming convention `lower-case-hypen`
For reference: https://www.openhab.org/docs/developer/guidelines.html#naming-convention

Author: fixed it

## Discovery

> Auto-discovery is supported and will discover all heatpumps within an account and also detect the specific channels supported by the model.

Reviewer: Or similar, just to make clear discovery depends on the bridge Thing. 
```suggestion
When the `account` bridge is setup, the binding will discover all heatpumps within that account and also detect the specific channels supported by the model.

```

## Bridge Configuration

> The following configuration parameters are available for the binding/bridge:

Reviewer: ```suggestion
The following configuration parameters are available for the bridge:
```

| clientSecret            | yes      | The secret which belongs to the clientId.                                                                                                                                                   |
| dataPollingInterval     | no       | Interval (seconds) in which live data values are retrieved from the Easee Cloud API. (default = 60)                                                                                         |

## Thing Configuration

It is recommended to use auto discovery which does not require further configuration.
If manual configuration is preferred you need to specify configuration as below.

| Configuration Parameter | Required | Description                                                                                                            |
|-------------------------|----------|------------------------------------------------------------------------------------------------------------------------|
| deviceId                | yes      | The id of the heatpump that will be represented by this thing. Can be retrieved via API call or autodiscovery.         |

> 

Reviewer: ```suggestion

```

## Channels

> The binding only supports all channels which are explicitely exposed by the myUplink API.

Reviewer: ```suggestion
The binding only supports channels which are explicitely exposed by the myUplink API.

```

Depending on your model and additional hardware the channels might be different. 

> Thus no list is provided here.

Reviewer: Have to look into the details, but in general it is best to show the channels. The above warning is already good enough for users to know not all channels would be available.

Author: I cannot provide any channels because the binding is generic and there are more than 50 different heat pumps (from different vendors) supported. They do not have a common set of channels. I would need to own one of each to add all the channels. In total this would be a list containing mor than 10.000 channels. I do not think that this would be useful for anybody.

Reviewer: No base set of common channels? I would have thought each heatpump would have a setpoint temperature for instance. Maybe not all, but just a few as example would already give some insights into the convention, structure that gives some guidance. If this also doesn’t make sense we will have to leave it as is.

Author: Unfortunately this is not the case. This is a multi-vendor API and although some channels like setpoint temperature exist for all models it will have different implementations e.g.
- different channel ID
- different min/max values
- different step values

Well this is the drawback of a generic API.

Reviewer: @jlaur or @lolodomo are you aware of any binding with a similar generic approach and how this is handled regading documentation?

## Full Example

_Provide a full usage example based on textual configuration files._
_*.things, *.items examples are mandatory as textual configuration is well used by many users._

> _*.sitemap examples are optional._

Reviewer: TODO

Author: documentation is updated. Basically the example was already there I just added a missing parameter and added a short explaination.

# myUplink Binding

The myUplink binding is used to get "live data" from from Nibe heat pumps without plugging any custom devices into your heat pump.
This avoids the risk of losing your warranty.
Instead data is retrieved from myUplink.
The myUplink API is the successor of the Nibe Uplink API.
This binding should in general be compatible with all heat pump models that support myUplink.
Read or write access is supported by all channels as exposed by the API.
Write access might only be available with a paid subscription for myUplink.

> You will need to create credentials at https://dev.myuplink.com/apps in order to use this binding.

Reviewer: ```suggestion
You will need to create credentials at <https://dev.myuplink.com/apps> in order to use this binding.
```

## Supported Things

This binding provides two thing types:

| Thing/Bridge        | Thing Type          | Description                                                                                   |
|---------------------|---------------------|-----------------------------------------------------------------------------------------------|
| bridge              | account             | cloud connection to a myUplink user account                                                   |

> | thing               | generic-device       | the physical heatpump which is connected to myUplink                                          |

Reviewer: nit: Please format markdown table

Author: done

> 

Reviewer: ```suggestion
```

## Discovery

When the `account` bridge is setup, the binding will discover all heatpumps within that account and also detect the specific channels supported by the model.

> 

Reviewer: ```suggestion
```

## Bridge Configuration

The following configuration parameters are available for the bridge:

| Configuration Parameter | Required | Description                                                                                                                                                                                 |
|-------------------------|----------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | clientId                | yes      | The clientId to login at myUplink cloud service. This is some kind of UUID. Visit https://dev.myuplink.com/apps to generate login credentials.                                              |

Reviewer: ```suggestion
| clientId                | yes      | The clientId to login at myUplink cloud service. This is some kind of UUID. Visit <https://dev.myuplink.com/apps> to generate login credentials.                                            |
```

| Configuration Parameter | Required | Description                                                                                                            |
|-------------------------|----------|------------------------------------------------------------------------------------------------------------------------|
| deviceId                | yes      | The id of the heatpump that will be represented by this thing. Can be retrieved via API call or autodiscovery.         |
| systemId                | no       | The systemId of the heatpump. Only needed for "SmartHomeMode". Can be retrieved via API call or autodiscovery.         |

## Channels

The binding only supports channels which are explicitely exposed by the myUplink API.

Depending on your model and additional hardware the channels might be different.
Thus no list is provided here.

## Full Example

The configuration below is an example which could easily be adopted to your actual model.
Thing configuration (account and generic-device) is the same for all models.
Item configuration depends on your specific model and thus channels will have different IDs and/or channels might not exist for all models.

> ### Thing Configuration

Reviewer: ```suggestion
### `demo.things` Example
```

```java
Bridge myuplink:account:myAccount "myUplink" [
    clientId="c7c2f9a4-b960-448f-b00d-b8f30aff3324",
    clientSecret="471147114711ABCDEF133713371337AB",
    dataPollingInterval=55
    ] {
        Thing generic-device vvm320 "VVM320" [ deviceId="id taken from automatic discovery", systemId="id taken from automatic discovery" ]
    }
```

> ### Item Configuration

Reviewer: ```suggestion
### `demo.items` Example
```
